gprofiler/metadata/cloud_metadata.py

```diff
@@ -0,0 +1,199 @@
+from typing import Dict, List, Optional
+
+import requests
+from requests import Response
+
+from gprofiler.exceptions import BadResponseCode
+from gprofiler.log import get_logger_adapter
+from gprofiler.metadata.metadata_type import Metadata
+
+METADATA_REQUEST_TIMEOUT = 5
+
+logger = get_logger_adapter(__name__)
+
+
+class InstanceMetadataBase:
+    def __init__(self, provider: str):
+        self.provider = provider
+
+
+class AwsInstanceMetadata(InstanceMetadataBase):
+    def __init__(
+        self,
+        provider: str,
+        region: str,
+        zone: str,
+        instance_type: str,
+        life_cycle: str,
+        account_id: str,
+        image_id: str,
+        instance_id: str,
+    ):
+        super().__init__(provider)
+        self.region = region
+        self.zone = zone
+        self.instance_type = instance_type
+        self.life_cycle = life_cycle
+        self.account_id = account_id
+        self.image_id = image_id
+        self.instance_id = instance_id
+
+
+class GcpInstanceMetadata(InstanceMetadataBase):
+    def __init__(
+        self,
+        provider: str,
+        zone: str,
+        instance_type: str,
+        preempted: bool,
+        preemptible: bool,
+        instance_id: str,
+        image_id: str,
+        name: str,
+    ):
+        super().__init__(provider)
+        self.zone = zone
+        self.instance_type = instance_type
+        self.preempted = preempted
+        self.preemptible = preemptible
+        self.instance_id = instance_id
+        self.image_id = image_id
+        self.name = name
+
+
+class AzureInstanceMetadata(InstanceMetadataBase):
+    def __init__(
+        self,
+        provider: str,
+        instance_type: str,
+        zone: str,
+        region: str,
+        subscription_id: str,
+        resource_group_name: str,
+        resource_id: str,
+        instance_id: str,
+        name: str,
+        image_info: Optional[Dict[str, str]],
+    ):
+        super().__init__(provider)
+        self.instance_type = instance_type
+        self.zone = zone
+        self.region = region
+        self.subscription_id = subscription_id
+        self.resource_group_name = resource_group_name
+        self.resource_id = resource_id
+        self.instance_id = instance_id
+        self.name = name
+        self.image_info = image_info
+
+
+def get_aws_metadata() -> Optional[AwsInstanceMetadata]:
+    # Documentation: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html
+    metadata_response = send_request("http://169.254.169.254/latest/dynamic/instance-identity/document")
```

Comment:
Shorter names, please make this function easier for the eyes. You can avoid extracting entries from the dict and just include them in the call: `AwsInstanceMetadata("aws", instance_metadata["region"], ...)`. Also, `instance_metadata` could be shorter. ...

Reply:
Please do so in the other providers as well.

Reply:
Shortened the entire file, now using dataclasses as well